Rubberduck.Parsing/VBA/ComponentParseTask.cs

```diff
@@ -0,0 +1,143 @@
+﻿using Antlr4.Runtime;
+using Antlr4.Runtime.Misc;
+using Antlr4.Runtime.Tree;
+using Microsoft.Vbe.Interop;
+using Rubberduck.Parsing.Grammar;
+using Rubberduck.Parsing.Nodes;
+using Rubberduck.Parsing.Preprocessing;
+using Rubberduck.Parsing.Symbols;
+using Rubberduck.VBEditor;
+using Rubberduck.VBEditor.Extensions;
+using System;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.Linq;
+using System.Runtime.InteropServices;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace Rubberduck.Parsing.VBA
+{
+    class ComponentParseTask
+    {
+        private readonly VBComponent _component;
+        private readonly QualifiedModuleName _qualifiedName;
+        private readonly TokenStreamRewriter _rewriter;
+        private readonly IAttributeParser _attributeParser;
+        private readonly VBAPreprocessor _preprocessor;
+
+        public event EventHandler<ParseCompletionArgs> ParseCompleted;
+        public event EventHandler<ParseFailureArgs> ParseFailure;
+
+        public ComponentParseTask(VBComponent vbComponent, VBAPreprocessor preprocessor, IAttributeParser attributeParser, TokenStreamRewriter rewriter = null)
+        {
+            _component = vbComponent;
+            _rewriter = rewriter;
+            _qualifiedName = new QualifiedModuleName(vbComponent); 
+        }
+
+        public Task ParseAsync(CancellationToken token)
+        {
+            return new Task(() => ParseInternal(token));
+        }
+
+        private void ParseInternal(CancellationToken token)
+        {
+            try
+            {
+                var code = RewriteAndPreprocess();
+                token.ThrowIfCancellationRequested();
+
+                var stopwatch = Stopwatch.StartNew();
+                ITokenStream stream;
+                var tree = ParseInternal(code, out stream);
+                stopwatch.Stop();
+                if (tree != null)
+                {
+                    Debug.Print("IParseTree for component '{0}' acquired in {1}ms (thread {2})", _component.Name, stopwatch.ElapsedMilliseconds, Thread.CurrentThread.ManagedThreadId);
+                }
+
+                token.ThrowIfCancellationRequested();
+
+                var attributes = _attributeParser.Parse(_component);
+
+                token.ThrowIfCancellationRequested();
+
+                ParseCompleted.Invoke(this, new ParseCompletionArgs
```

Comment:
Invoking the event without first copying it to a local field means this is not thread-safe which is particulary relevant since the action you're doing here is inherently multithreaded. This goes for all event invocations you're doing I see from a quick scan.